src/main/scala/mesosphere/marathon/core/storage/repository/impl/TemplateRepository.scala

```diff
@@ -0,0 +1,174 @@
+import mesosphere.marathon.core.storage.zookeeper.ZooKeeperPersistenceStore
+import mesosphere.marathon.state._
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.hashing.MurmurHash3
+import scala.util.{Failure, Success}
+
+/**
+  * This class implements a repository for templates. It uses [[BucketingRepository]] (and underlying [[ZooKeeperPersistenceStore]]
+  * to store templates and their versions in a tree like structure in the Zookeeper.
+  *
+  * Using [[BucketingRepository]] implementation, all template versions are saved like `/$base/$bucket/$safePathId/$version` where:
+  * <ul>
+  * <li>`$base` is simply "/templates"</li>
+  * <li>`$bucket` is bucket hash of templates pathId (see [[mesosphere.marathon.core.storage.repository.impl.TemplateRepository.TemplateBucketPath]]
+  *   implementation for details)</li>
+  * <li> `$safePathId` is [[PathId.safePath]] (see [[mesosphere.marathon.core.storage.repository.impl.TemplateRepository.VersionBucketPath]]
+  *   implementation for details)</li>
+  * <li>`$version` is template version</li>
+  * </ul>
+  *
+  * So two templates with [[PathId]]s `/eng/foo` (with versions 1 and 2) and `/sales/bar` (versions 3) will be stored like following:
+  * {{{
+  * /templates
+  *   /01
+  *     /eng_foo
+  *       /1
+  *       /2
+  *   /05
+  *     /sales_bar
+  *       /3
+  * }}}
+  * Most methods simply use the underlying [[BucketingRepository]] methods to store templates by mapping from [[PathId]] and version
+  * to [[BucketNode]] and [[BucketPath]] using [[mesosphere.marathon.core.storage.repository.impl.TemplateRepository.VersionBucketPath]]
+  * and [[mesosphere.marathon.core.storage.repository.impl.TemplateRepository.TemplateBucketPath]] helper classes.
+  *
+  * [[create]] method takes a [[Spec]] and returns a [[Template]] which contains the spec and it's version. A version is an integer that
+  * is automatically generated and assigned by the repository. Internally this repository holds a counter for every template incrementing
+  * it every time a new template version is created.
+  *
+  * [[update()]] method is overridden and throws an [[UnsupportedOperationException]] since templates are immutable and should not
+  * be updated.
+  *
+  * [[init()]] method initializes repository by reading all existing templates Ids and their versions (but not template content) and
+  * initializing internal template version counters. This method should be called and waited for successful competition prior to any
+  * other usage of the repository.
+  *
+  * Known pitfalls: please note that while there is one at least one version of the template, a counter is always increased monotonically.
+  * However should all template versions be deleted and marathon restarted, a counter is initialized with `0`.
+  * This implementation is tread-safe and provides serializability guarantees e.g. two independent [[create()]] calls are not guaranteed
```

Comment:
But two serial creates are run serially, right?

Reply:
BTW what width are you using? I'm showing... 127-130?

Reply:
> But two serial creates are run serially, in order, right?

Well, yes, when called from the same thread it will because of the lock-free implementation using the atomic counter:
```
// In order:
Thread-1:
repo.create(/foo) <- version 1
repo.create(/foo) <- version 2
```

However, I wanted to give the same guaranties for `TemplateRepository` and it's cached version `CachedTemplateRepository` where:
```
//No order:
async {
  cachedRepo.create(/foo) <-- version 1 or 2
  cachedRepo.create(/foo) <-- version 1 or 2
}

// In order again with await
async {
  await(cachedRepo.create(/foo)) <-- version 1
  await(cachedRepo.create(/foo)) <-- version 2
}
```